### PR TITLE
LB-1630: Dashboard for users with sporadic listens loads slowly

### PR DESCRIPTION
**Pull request by fettuccinae**
# Problem

Slow lookup times for users with sporadic listens. [Ticket.](https://tickets.metabrainz.org/browse/LB-1630)

# Solution

1. Changed the maximum number of passes from 10 to 3, fixing the maximum size of search window to 9 months in one pass.
2. Api returns search_start_ts and search_end_ts so that frontend uses these timestamps and doesnt make redundant requests.
3. If listens are less than expected, render a Load More listens button which uses the search_end_ts as max_ts and makes next request.

### CODE REVIEW

**Comment by fettuccinae** [comment]
@amCap1712 I have modified the fetch_listens method as suggested. If we limit the passes itself, we dont need to do anything different for initial load.

**Comment by fettuccinae** [comment]
@amCap1712

File: listenbrainz/listenstore/timescale_listenstore.py [hunk 1]
```diff
@@ -365,14 +347,30 @@ def fetch_listens(self, user: Dict, from_ts: datetime = None, to_ts: datetime =
...
+                    window_size *= WINDOW_SIZE_MULTIPLIER
+                    to_ts += window_size
+
+                if from_dynamic:
+                    to_ts -= window_size
+                    window_size *= WINDOW_SIZE_MULTIPLIER
+                    from_ts -= window_size
+
+        # Search ends at this timestamp.
+        dynamic_ts = from_ts if from_dynamic else to_ts
```
Comment (amCap1712) [inline listenbrainz/listenstore/timescale_listenstore.py:363]: Consider the case where the window was extended and a number of listens greater than the limit were found, in that case the while loop above would terminate before exhausting the search timestamps and the caller of this function would skip the timestamps to the end of the window while actually it should start from the the timestamp of the last used listen.

So you should take into consideration the min/max timestamp of the listens returned and not just the window limits.

**Review by amCap1712 — CHANGES_REQUESTED** [review]
I simplified the logic in updating the window sizes but there is a bug that needs to be resolved.

File: listenbrainz/listenstore/timescale_listenstore.py (hunk 1 above)
Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:363]: Makes sense but this information will be eventually exposed through the API as well so it makes sense to still change it for avoiding confusion with other clients.

Comment (fettuccinae) [reply listenbrainz/listenstore/timescale_listenstore.py:363]: Ahh understood, so , if len(listens) >= limit, set the search_end_ts as last listen's timestamp?

Comment (amCap1712) [reply listenbrainz/listenstore/timescale_listenstore.py:363]: Off the top of my head yes but also take into consideration the order of sorting, from_ts and to_ts. And add tests as well for the same. I'll need to think more on this to see if there are any other edge cases.

**Comment by amCap1712** [comment]
Hi! The code works fine but I haven't merged it yet because I am thinking about the changes in behavior for the API users.

**Comment by fettuccinae** [comment]
> Hi! The code works fine but I haven't merged it yet because I am thinking about the changes in behavior for the API users.

Hey! For the api users, we can keep calling the fetch_listens function till 25 listens are reached or till the search_end_ts has exceeded min_user_ts, or we can just change the api description to "fetching upto 25 listens in 13 months from the timestamp".

**Comment by MonkeyDo** [comment]
@amCap1712 As far as i understand, the website part of this PR should be ready for review?
Let me know if there are more changes you want to implement regarding LB web page loading, will review otherwise.

**Comment by amCap1712** [comment]
@MonkeyDo I discussed this with @mayhem on IRC and we are not entirely happy with this behavior for the API side of things.
